recipes-app/switchserialmode/files/src/switchserialmode.h

```diff
@@ -0,0 +1,45 @@
+#ifndef SWITCH_SERIAL_MODE
+#define SWITCH_SERIAL_MODE
+
+#include <stdint.h>
+
+#define ERROR       -1
+#define SUCCESS     0
+
+typedef enum boardType {
+    BASIC_BOARD = 0,
+    ADVANCED_BOARD,
+    ADVANCED_BOARD_PG1
+} boardType_e;
+
+typedef struct transceiver_ops {
+    void (*transceiver_set_termination)( uint8_t onoff);
+    void (*transceiver_switch_mode)(const uint8_t *mode);
+} transceiver_ops_t;
+
+typedef struct {
+    /*Device node*/
+    uint8_t *devName;
+
+    /*Init device resources*/
+    int8_t (*init)(uint8_t *devName);
+
+    /*Set serial mode,like rs232 rs485 rs422*/
+    void (*setMode)(uint8_t *mode);
+
+    /*Display current serial mode*/
+    void (*getMode)(void);
+
+    /*For the device which support setup and hold time*/
+    void (*rs485HoldTime)(uint8_t time);
+    void (*rs485SetupTime)(uint8_t time);
+
+    /*Release*/
+    void (*release)(void);
+
+    transceiver_ops_t *transOps;
+}serial_ops_t;
```

Comment:
Better to have a platform.c to include the platform definitions, and put serial and transceiver at the same level, below platform, as element of device arrays.

Reply:
For example:

```
struct device {
    const char *name;
    serial_ops_t *ctrl_ops;
    transceiver_ops_t *xcvr_ops;
};

struct platform {
    const char *name;
    uint8_t ndev;
    struct device [];
};

struct platform platforms[] = {
    {"IOT2050-BASIC", 1, {"ttyX30", &omap_8250_ops. &sp339eOps}};
    ...
};

```

Reply:
This gives you the most flexibility to support more platform, and support more devices.

Reply:
now I use board type to initial all the ops at runtime.
if we need to support more boards, I think we can mark a TBD to do in the future.